metricproviders/datadog/datadog.go

```diff
@@ -0,0 +1,147 @@
+package datadog
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strconv"
+	"time"
+
+	metricutil "github.com/argoproj/argo-rollouts/utils/metric"
+	log "github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/util/jsonpath"
+
+	"github.com/argoproj/argo-rollouts/pkg/apis/rollouts/v1alpha1"
+	"github.com/argoproj/argo-rollouts/utils/evaluate"
+)
+
+var unixNow = func() int64 { return time.Now().Unix() }
+
+const (
+	//ProviderType indicates the provider is datadog
+	ProviderType = "Datadog"
+)
+
+// Provider contains all the required components to run a Datadog query
+// Implements the Provider Interface
+type Provider struct {
+	logCtx log.Entry
+}
+
+// Type incidates provider is a Datadog provider
+func (p *Provider) Type() string {
+	return ProviderType
+}
+
+func (p *Provider) Run(run *v1alpha1.AnalysisRun, metric v1alpha1.Metric) v1alpha1.Measurement {
+	startTime := metav1.Now()
+
+	// Measurement to pass back
+	measurement := v1alpha1.Measurement{
+		StartedAt: &startTime,
+	}
+
+	endpoint := "https://api.datadoghq.com/api/v1/query"
+	if metric.Provider.Datadog.URL != "" {
+		endpoint = metric.Provider.Datadog.URL + "/api/v1/query"
+	}
+
+	url, _ := url.Parse(endpoint)
+
+	now := unixNow()
+	q := url.Query()
+	q.Set("query", metric.Provider.Datadog.Query)
+	q.Set("from", strconv.FormatInt(now-300, 10))
+	q.Set("to", strconv.FormatInt(now, 10))
```

Comment:
The 300 value (5 minutes) used in the `from` is a hard wired value that probably won't work for everyone. I think the interval needs to be user specified in the spec as an `interval` duration string field. e.g.:

```yaml
    provider:
      datadog:
        interval: 5m
        URL: https://api.datadoghq.com
        query: |
          sum:requests.error.count{service:{{args.service-name}}} /
          sum:requests.request.count{service:{{args.service-name}}}

```